### PR TITLE
Sample using AppOnly Authentication

### PR DESCRIPTION
**Pull request by OfirGordon**
Added support for AppOnly authentication for AppOwnsData samples - embed report, dashboard and tile.
Split config file to 2 sections - one for each authentication method,
and added support in HomeController.

### CODE REVIEW

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config [hunk 1]
```diff
@@ -7,20 +7,54 @@
...
     <add key="pbiPassword" value=""/>
-  </appSettings>
+
+    <add key="authorityUrl" value="https://login.windows.net/common/oauth2/authorize/" />
+    <add key="resourceUrl" value="https://analysis.windows.net/powerbi/api" />
+    <add key="apiUrl" value="https://api.powerbi.com/" />
+    <add key="embedUrlBase" value="https://app.powerbi.com/" />
+  </MasterUser>
+
+  <ServicePrinciple>
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config]: ServicePrincipal

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config [hunk 2]
```diff
@@ -7,20 +7,54 @@
...
+    <add key="applicationId" value="" />
+    <!-- Note: Do NOT leave your app secret on code. Save it in secure place. -->
+    <add key="applicationSecret" value="" />
+
+    <add key="workspaceId" value=""/>
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config]: why do we need to duplicate all of the config entries.
the only difference is whether the user credentials are needed or not.

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 3]
```diff
@@ -139,146 +134,139 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
...
-                if (tokenResponse == null)
-                {
-                    return View(new EmbedConfig()
+                    if (tokenResponse == null)
                     {
-                        ErrorMessage = "Failed to generate embed token."
-                    });
-                }
+                        result.ErrorMessage = "Failed to generate embed token.";
+                        return View(result);
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: unnecessary change

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 4]
```diff
@@ -139,146 +134,139 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
...
-                if (dashboard == null)
-                {
-                    return View(new EmbedConfig()
+                    if (dashboard == null)
                     {
-                        ErrorMessage = "Workspace has no dashboards."
-                    });
-                }
+                        result.ErrorMessage = "Workspace has no dashboards.";
+                        return View(result);
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: unnecessary change

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 5]
```diff
@@ -139,146 +134,139 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
...
-                    EmbedUrl = dashboard.EmbedUrl,
-                    Id = dashboard.Id
-                };
+                    // Generate Embed Configuration.
+                    result.EmbedToken = tokenResponse;
+                    result.EmbedUrl = dashboard.EmbedUrl;
+                    result.Id = dashboard.Id;
 
-                return View(embedConfig);
+                    return View(result);
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: here as well

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 6]
```diff
@@ -139,146 +134,139 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
...
+                }
             }
+            catch (HttpOperationException exc)
+            {
+                result.ErrorMessage = string.Format("Status: {0} ({1})\r\nResponse: {2}\r\nRequestId: {3}", exc.Response.StatusCode, (int)exc.Response.StatusCode, exc.Response.Content, exc.Response.Headers["RequestId"].FirstOrDefault());
+            }
+            catch (Exception exc)
+            {
+                result.ErrorMessage = exc.ToString();
```
Comment (MahirDiab) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: do not catch general exception.
browser handling will be more informative than "exc.ToString()"

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config (hunk 2 above)
Comment (MahirDiab) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config]: i suggest we have one common config section with all required metadata for both scenarios which includes "IdentutyType = DelegatedUser | App" and 2 config sections 
App: clientSecret
DelegatedUser: username & password

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs (hunk 4 above)
Comment (OfirGordon) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: I think it's better if it is the same in all embed functions

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs (hunk 6 above)
Comment (OfirGordon) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: it's in addition to catching HttpOperationException and handling all relevant errors in the function scope (assigning the error to result.ErrorMessage and printing to browser).
this catch is just for any other case...

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 7]
```diff
@@ -77,7 +77,7 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
                     }
                     else
                     {
-                        report = reports.Value.FirstOrDefault(r => r.Id == ReportId);
+                        report = reports.Value.FirstOrDefault(r => r.Id.Equals(ReportId, StringComparison.InvariantCultureIgnoreCase));
```
Comment (may-hartov) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: Why do we need this casing comparison?

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs (hunk 6 above)
Comment (may-hartov) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: Wrap with try only the statements that may result in exception, and catch only the corresponding exceptions. 
It is easier to debug, and you can create an informative error message for each case when you know what can fail and why

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 8]
```diff
@@ -139,146 +139,145 @@ public async Task<ActionResult> EmbedReport(string username, string roles)
 
         public async Task<ActionResult> EmbedDashboard()
         {
-            var error = GetWebConfigErrors();
-            if (error != null)
+            var result = new EmbedConfig();
+            try
             {
-                return View(new EmbedConfig()
+                var error = GetWebConfigErrors();
```
Comment (may-hartov) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: I don't think this should be in the try block. If any exception can be thrown in GetWebConfigErrors, catch and return error from within the message, although it seems to me that GetWebConfigErrors doesn't get errors

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs [hunk 9]
```diff
@@ -14,21 +15,25 @@ namespace PowerBIEmbedded_AppOwnsData.Controllers
 {
     public class HomeController : Controller
```
Comment (may-hartov) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: ReportEmbed, TileEmbed and DashboardEmbd() share a lot of code, and they are very long. Maybe you can extract some of the common code and as a result have the methods smaller and avoid duplications

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs (hunk 7 above)
Comment (OfirGordon) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: In some cases the returning string contains cap letters, without this comparison we'll get an empty result

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs (hunk 9 above)
Comment (OfirGordon) [reply App Owns Data/PowerBIEmbedded_AppOwnsData/Controllers/HomeController.cs]: Most of the shared code is in the authentication part and I tried to merged it as much as possible, I think that trying to reduce it more will result in an unclear code.
In addition to that, this task purpose is to add the AppOnly authentication feature, and there is another major fix upcoming which will include more reduction for the shared code.

File: App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config [hunk 10]
```diff
@@ -7,20 +7,40 @@
...
+  </appSettings>
+
+  <MasterUser>
     <!-- Note: Do NOT leave your credentials on code. Save them in secure place. -->
     <add key="pbiUsername" value=""/>
     <add key="pbiPassword" value=""/>
-  </appSettings>
+  </MasterUser>
+
+  <ServicePrincipal>
```
Comment (may-hartov) [inline App Owns Data/PowerBIEmbedded_AppOwnsData/Web.config]: Add a comment for MasterUser vs ServicePrincipal usage
